clippy_lints/src/unintentional_unit_return.rs

```diff
@@ -0,0 +1,155 @@
+    /// way of specifying this without triggering needless_return lint
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let mut twins = vec!((1,1), (2,2));
+    /// twins.sort_by_key(|x| { x.1; });
+    /// ```
+    pub UNINTENTIONAL_UNIT_RETURN,
+    nursery,
+    "fn arguments of type Fn(...) -> Once having last statements with a semi-colon, suggesting to remove the semi-colon if it is unintentional."
+}
+
+declare_lint_pass!(UnintentionalUnitReturn => [UNINTENTIONAL_UNIT_RETURN]);
+
+fn unwrap_trait_pred<'tcx>(cx: &LateContext<'_, 'tcx>, pred: &Predicate<'tcx>) -> Option<TraitPredicate<'tcx>> {
+    if let Predicate::Trait(poly_trait_pred, _) = pred {
+        let trait_pred = cx.tcx.erase_late_bound_regions(&poly_trait_pred);
+        Some(trait_pred)
+    } else {
+        None
+    }
+}
+
+fn get_predicates_for_trait_path<'tcx>(
+    cx: &LateContext<'_, 'tcx>,
+    generics: GenericPredicates<'tcx>,
+    trait_path: &[&str],
+) -> Vec<&'tcx Predicate<'tcx>> {
+    let mut preds = Vec::new();
+    generics.predicates.iter().for_each(|(pred, _)| {
+        if_chain! {
+            if let Some(trait_pred) = unwrap_trait_pred(cx, pred);
+            if let Some(trait_def_id) = get_trait_def_id(cx, trait_path);
+            if trait_def_id == trait_pred.trait_ref.def_id;
+            then {
+                    preds.push(pred);
+                }
+        }
+    });
+    preds
+}
+
+fn get_projection_pred<'tcx>(
+    cx: &LateContext<'_, 'tcx>,
+    generics: GenericPredicates<'tcx>,
+    pred: TraitPredicate<'tcx>,
+) -> Option<ProjectionPredicate<'tcx>> {
+    generics.predicates.iter().find_map(|(proj_pred, _)| {
+        if let Predicate::Projection(proj_pred) = proj_pred {
+            let projection_pred = cx.tcx.erase_late_bound_regions(proj_pred);
+            if projection_pred.projection_ty.substs == pred.trait_ref.substs {
+                return Some(projection_pred);
+            }
+        }
+        None
+    })
+}
+
+fn get_args_to_check<'tcx>(cx: &LateContext<'_, 'tcx>, expr: &'tcx Expr<'tcx>) -> Vec<(usize, String)> {
+    let mut args_to_check = Vec::new();
+    if let Some(def_id) = cx.tables.type_dependent_def_id(expr.hir_id) {
+        let fn_sig = cx.tcx.fn_sig(def_id);
+        let generics = cx.tcx.predicates_of(def_id);
+        let fn_mut_preds = get_predicates_for_trait_path(cx, generics, &paths::FN_MUT);
+        let ord_preds = get_predicates_for_trait_path(cx, generics, &paths::ORD);
+        let partial_ord_preds = get_predicates_for_trait_path(cx, generics, &paths::PARTIAL_ORD);
+        // Trying to call erase_late_bound_regions on fn_sig.inputs() gives the following error
+        // The trait `rustc::ty::TypeFoldable<'_>` is not implemented for `&[&rustc::ty::TyS<'_>]`
+        let inputs_output = cx.tcx.erase_late_bound_regions(&fn_sig.inputs_and_output());
+        inputs_output.iter().enumerate().for_each(|(i, inp)| {
```

Comment:
This can be done with 
```suggestion
        inputs_output.iter().rev().skip(1).rev().enumerate().for_each(|(i, inp)| {
```